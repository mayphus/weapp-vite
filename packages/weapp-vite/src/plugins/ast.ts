import type { Node } from 'estree'
// import type { ProgramNode } from 'rollup'
import { walk } from 'estree-walker'

export function collectRequireTokens(ast: Node) {
  const requireTokens: string[] = []
  const importExpressions: {
    start: number
    end: number
    value: string
  }[] = []
  walk(ast, {
    enter(node) {
      if (node.type === 'CallExpression') {
        if (node.callee.type === 'Identifier' && node.callee.name === 'require') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireTokens.push(node.arguments[0].value)
          }
        }
        else if (node.callee.type === 'MemberExpression' && node.callee.object.type === 'Identifier' && node.callee.object.name === 'require' && node.callee.property.type === 'Identifier' && node.callee.property.name === 'async') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireTokens.push(node.arguments[0].value)
          }
        }
      }
      else if (node.type === 'ImportExpression') {
        importExpressions.push({
          // @ts-ignore
          start: node.start,
          // @ts-ignore
          end: node.start + 'import'.length,
          value: 'require.async',
        })

        if (node.source.type === 'Literal' && typeof node.source.value === 'string') {
          requireTokens.push(node.source.value)
        }
      }
    },
  })

  return {
    requireTokens,
    importExpressions,
  }
}
