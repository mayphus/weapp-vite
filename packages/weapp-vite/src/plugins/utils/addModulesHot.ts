import type { PluginContext } from 'rollup'

export function addModulesHot(entriesSet: Set<string>, pluginContext: PluginContext) {
  for (const entry of entriesSet) {
    const moduleInfo = pluginContext.getModuleInfo(entry)

    if (moduleInfo) {
      const stack = [moduleInfo.id] // 用栈模拟递归
      const visitedModules = new Set<string>()

      while (stack.length > 0) {
        const id = stack.pop()

        if (id && !visitedModules.has(id)) {
          visitedModules.add(id)

          const info = pluginContext.getModuleInfo(id)

          if (info) {
            pluginContext.addWatchFile(info.id)
            // 将子依赖加入栈
            stack.push(...info.importedIds)
          }
        }
      }
    }
  }
}
