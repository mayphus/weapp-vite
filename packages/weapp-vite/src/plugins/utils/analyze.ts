import type { App, Component, Page } from '@weapp-core/schematics'
import { get } from '@weapp-core/shared'

export function analyzeAppJson(json: App) {
  const entries: string[] = []
  const pages = json.pages ?? []
  const components = Object.values(json.usingComponents ?? {})
  const subPackages = (
    [...json.subPackages ?? [], ...json.subpackages ?? []]?.reduce<string[]>(
      (acc, cur) => {
        acc.push(...(cur.pages ?? []).map(x => `${cur.root}/${x}`))
        return acc
      },
      [],
    )) ?? []

  if (get(json, 'tabBar.custom')) {
    entries.push('custom-tab-bar/index')
  }
  // 全局工具栏
  // https://developers.weixin.qq.com/miniprogram/dev/framework/runtime/skyline/appbar.html
  if (get(json, 'appBar')) {
    entries.push('app-bar/index')
  }
  entries.push(
    ...pages,
    ...components,
    ...subPackages,
  )
  return entries
}

export function analyzeCommonJson(json: Page | Component) {
  const entries: string[] = []
  const components = Object.values(json.usingComponents ?? {})
  entries.push(...components)
  return entries
}
