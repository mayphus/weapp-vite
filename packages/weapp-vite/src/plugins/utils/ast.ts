import type { Node } from 'estree'
// import type { ProgramNode } from 'rollup'
import { walk } from 'estree-walker'

export function collectRequireTokens(ast: Node) {
  const requireModules: {
    start: number
    end: number
    value: string
    async?: boolean
    // leadingComment: string
  }[] = []
  walk(ast, {
    enter(node) {
      if (node.type === 'CallExpression') {
        // TODO 可能要被废弃
        // https://developers.weixin.qq.com/miniprogram/dev/reference/api/require.html#require-async-%E9%93%BE%E5%BC%8F%E8%B0%83%E7%94%A8
        if (node.callee.type === 'Identifier' && node.callee.name === 'require') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireModules.push({
              // @ts-ignore
              start: node.arguments[0].start,
              // @ts-ignore
              end: node.arguments[0].end,
              value: node.arguments[0].value,
            })
          }
        }
        else if (
          node.callee.type === 'MemberExpression'
          && node.callee.object.type === 'Identifier'
          && node.callee.object.name === 'require'
          && node.callee.property.type === 'Identifier'
          && node.callee.property.name === 'async') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireModules.push({
              // @ts-ignore
              start: node.arguments[0].start,
              // @ts-ignore
              end: node.arguments[0].end,
              value: node.arguments[0].value,
              async: true,
            })
          }
        }
      }
    },
  })

  return {
    requireModules,
  }
}
