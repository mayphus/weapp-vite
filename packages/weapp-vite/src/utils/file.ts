import { configExtensions, jsExtensions, templateExtensions, vueExtensions } from '@/constants'
import fs from 'fs-extra'
import path from 'pathe'

export function isJsOrTs(name?: string) {
  if (typeof name === 'string') {
    return jsExtensions.some(x => name.endsWith(`.${x}`))
  }
  return false
}

export function isTemplateRequest(request: string) {
  return request.endsWith('.wxml') || request.endsWith('.html')
}

export function changeFileExtension(filePath: string, extension: string) {
  if (typeof filePath !== 'string') {
    throw new TypeError(`Expected \`filePath\` to be a string, got \`${typeof filePath}\`.`)
  }

  if (typeof extension !== 'string') {
    throw new TypeError(`Expected \`extension\` to be a string, got \`${typeof extension}\`.`)
  }

  if (filePath === '') {
    return ''
  }

  extension = extension ? (extension.startsWith('.') ? extension : `.${extension}`) : ''

  const basename = path.basename(filePath, path.extname(filePath))
  return path.join(path.dirname(filePath), basename + extension)
}

export async function findVueEntry(filepath: string) {
  for (const ext of vueExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsEntry(filepath: string) {
  for (const ext of jsExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsonEntry(filepath: string) {
  for (const ext of configExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findTemplateEntry(filepath: string) {
  for (const ext of templateExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export function isTemplate(filepath: string) {
  return templateExtensions.some(ext => filepath.endsWith(`.${ext}`))
}

export function touch(filename: string) {
  const time = new Date()

  try {
    fs.utimesSync(filename, time, time)
  }
  catch {
    fs.closeSync(fs.openSync(filename, 'w'))
  }
}
