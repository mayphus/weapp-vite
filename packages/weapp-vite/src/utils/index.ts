export * from './file'
export * from './json'
export * from './projectConfig'
export * from './scan'
// export * from './wxml'
// export * from './wxs'
