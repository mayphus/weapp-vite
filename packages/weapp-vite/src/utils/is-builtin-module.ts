import builtinModules from './builtin-modules.json'

let moduleSet
export default function isBuiltinModule(moduleName: string) {
  if (typeof moduleName !== 'string') {
    throw new TypeError('Expected a string')
  }

  moduleSet ??= new Set(builtinModules)

  return moduleSet.has(moduleName)
}
