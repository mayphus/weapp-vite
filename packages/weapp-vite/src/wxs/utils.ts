import { addExtension, removeExtensionDeep } from '@weapp-core/shared'

export function normalizeWxsFilename(value: string) {
  return addExtension(removeExtensionDeep(value), '.wxs')
}
