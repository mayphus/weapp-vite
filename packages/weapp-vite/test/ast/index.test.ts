import fs from 'fs-extra'
import MagicString from 'magic-string'
import path from 'pathe'
import { parseAst, parseAstAsync } from 'rollup/parseAst'
import { collectRequireTokens } from '@/plugins/utils/ast'

function normalizeCode(code: string) {
  return code.replace(/\r\n/g, '\n')
}

describe('require', () => {
  it('should 0', async () => {
    const code = normalizeCode(await fs.readFile(path.resolve(__dirname, './fixtures/require/index.ts'), 'utf-8'))
    const ast = parseAst(code)
    expect(ast).toMatchSnapshot()
  })

  it('should 1', async () => {
    const code = normalizeCode(await fs.readFile(path.resolve(__dirname, './fixtures/require/index.ts'), 'utf-8'))
    const ast = await parseAstAsync(code)
    const ms = new MagicString(code)
    const { requireModules } = collectRequireTokens(ast)

    expect(requireModules).toMatchSnapshot()
    expect(ms.toString()).toMatchSnapshot()
    for (const m of requireModules) {
      const x = ms.slice(m.start, m.end)
      expect(x).toMatchSnapshot()
    }
  })

  it('case0.js', async () => {
    const code = normalizeCode(await fs.readFile(path.resolve(__dirname, './fixtures/case0.js'), 'utf-8'))
    const ast = await parseAstAsync(code)
    const { requireModules } = collectRequireTokens(ast)
    expect(requireModules).toMatchSnapshot('requireModules')
  })

  it('case1.js', async () => {
    const code = normalizeCode(await fs.readFile(path.resolve(__dirname, './fixtures/case1.js'), 'utf-8'))
    const ast = await parseAstAsync(code)
    const { requireModules } = collectRequireTokens(ast)
    expect(requireModules).toMatchSnapshot('requireModules')
  })

  it('case2.js', async () => {
    const code = normalizeCode(await fs.readFile(path.resolve(__dirname, './fixtures/case2.js'), 'utf-8'))
    const ast = await parseAstAsync(code)
    const { requireModules } = collectRequireTokens(ast)
    expect(requireModules).toMatchSnapshot('requireModules')
  })
})
