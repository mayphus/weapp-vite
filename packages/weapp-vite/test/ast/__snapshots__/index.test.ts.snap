// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`require > should 0 1`] = `
{
  "body": [
    {
      "declarations": [
        {
          "end": 34,
          "id": {
            "end": 14,
            "properties": [
              {
                "computed": false,
                "end": 12,
                "key": {
                  "end": 12,
                  "name": "part",
                  "start": 8,
                  "type": "Identifier",
                },
                "kind": "init",
                "method": false,
                "shorthand": true,
                "start": 8,
                "type": "Property",
                "value": {
                  "end": 12,
                  "name": "part",
                  "start": 8,
                  "type": "Identifier",
                },
              },
            ],
            "start": 6,
            "type": "ObjectPattern",
          },
          "init": {
            "arguments": [
              {
                "end": 33,
                "raw": "'./part'",
                "start": 25,
                "type": "Literal",
                "value": "./part",
              },
            ],
            "callee": {
              "end": 24,
              "name": "require",
              "start": 17,
              "type": "Identifier",
            },
            "end": 34,
            "optional": false,
            "start": 17,
            "type": "CallExpression",
          },
          "start": 6,
          "type": "VariableDeclarator",
        },
      ],
      "end": 34,
      "kind": "const",
      "start": 0,
      "type": "VariableDeclaration",
    },
    {
      "declarations": [
        {
          "end": 72,
          "id": {
            "end": 51,
            "properties": [
              {
                "computed": false,
                "end": 49,
                "key": {
                  "end": 49,
                  "name": "other",
                  "start": 44,
                  "type": "Identifier",
                },
                "kind": "init",
                "method": false,
                "shorthand": true,
                "start": 44,
                "type": "Property",
                "value": {
                  "end": 49,
                  "name": "other",
                  "start": 44,
                  "type": "Identifier",
                },
              },
            ],
            "start": 42,
            "type": "ObjectPattern",
          },
          "init": {
            "arguments": [
              {
                "end": 71,
                "raw": "'./other'",
                "start": 62,
                "type": "Literal",
                "value": "./other",
              },
            ],
            "callee": {
              "end": 61,
              "name": "require",
              "start": 54,
              "type": "Identifier",
            },
            "end": 72,
            "optional": false,
            "start": 54,
            "type": "CallExpression",
          },
          "start": 42,
          "type": "VariableDeclarator",
        },
      ],
      "end": 72,
      "kind": "const",
      "start": 36,
      "type": "VariableDeclaration",
    },
    {
      "end": 249,
      "expression": {
        "arguments": [
          {
            "end": 248,
            "properties": [
              {
                "computed": false,
                "end": 245,
                "key": {
                  "end": 95,
                  "name": "onLoad",
                  "start": 89,
                  "type": "Identifier",
                },
                "kind": "init",
                "method": true,
                "shorthand": false,
                "start": 83,
                "type": "Property",
                "value": {
                  "async": true,
                  "body": {
                    "body": [
                      {
                        "declarations": [
                          {
                            "end": 153,
                            "id": {
                              "end": 113,
                              "name": "mod",
                              "start": 110,
                              "type": "Identifier",
                            },
                            "init": {
                              "argument": {
                                "arguments": [
                                  {
                                    "end": 152,
                                    "raw": "'./async'",
                                    "start": 143,
                                    "type": "Literal",
                                    "value": "./async",
                                  },
                                ],
                                "callee": {
                                  "computed": false,
                                  "end": 142,
                                  "object": {
                                    "end": 129,
                                    "name": "require",
                                    "start": 122,
                                    "type": "Identifier",
                                  },
                                  "optional": false,
                                  "property": {
                                    "end": 142,
                                    "name": "async",
                                    "start": 137,
                                    "type": "Identifier",
                                  },
                                  "start": 122,
                                  "type": "MemberExpression",
                                },
                                "end": 153,
                                "optional": false,
                                "start": 122,
                                "type": "CallExpression",
                              },
                              "end": 153,
                              "start": 116,
                              "type": "AwaitExpression",
                            },
                            "start": 110,
                            "type": "VariableDeclarator",
                          },
                        ],
                        "end": 153,
                        "kind": "const",
                        "start": 104,
                        "type": "VariableDeclaration",
                      },
                      {
                        "declarations": [
                          {
                            "end": 200,
                            "id": {
                              "end": 173,
                              "name": "mod2",
                              "start": 169,
                              "type": "Identifier",
                            },
                            "init": {
                              "argument": {
                                "end": 200,
                                "options": null,
                                "source": {
                                  "end": 199,
                                  "raw": "'./async2'",
                                  "start": 189,
                                  "type": "Literal",
                                  "value": "./async2",
                                },
                                "start": 182,
                                "type": "ImportExpression",
                              },
                              "end": 200,
                              "start": 176,
                              "type": "AwaitExpression",
                            },
                            "start": 169,
                            "type": "VariableDeclarator",
                          },
                        ],
                        "end": 200,
                        "kind": "const",
                        "start": 163,
                        "type": "VariableDeclaration",
                      },
                      {
                        "end": 241,
                        "expression": {
                          "arguments": [
                            {
                              "end": 222,
                              "name": "part",
                              "start": 218,
                              "type": "Identifier",
                            },
                            {
                              "end": 229,
                              "name": "other",
                              "start": 224,
                              "type": "Identifier",
                            },
                            {
                              "end": 234,
                              "name": "mod",
                              "start": 231,
                              "type": "Identifier",
                            },
                            {
                              "end": 240,
                              "name": "mod2",
                              "start": 236,
                              "type": "Identifier",
                            },
                          ],
                          "callee": {
                            "computed": false,
                            "end": 217,
                            "object": {
                              "end": 213,
                              "name": "console",
                              "start": 206,
                              "type": "Identifier",
                            },
                            "optional": false,
                            "property": {
                              "end": 217,
                              "name": "log",
                              "start": 214,
                              "type": "Identifier",
                            },
                            "start": 206,
                            "type": "MemberExpression",
                          },
                          "end": 241,
                          "optional": false,
                          "start": 206,
                          "type": "CallExpression",
                        },
                        "start": 206,
                        "type": "ExpressionStatement",
                      },
                    ],
                    "end": 245,
                    "start": 98,
                    "type": "BlockStatement",
                  },
                  "end": 245,
                  "expression": false,
                  "generator": false,
                  "id": null,
                  "params": [],
                  "start": 95,
                  "type": "FunctionExpression",
                },
              },
            ],
            "start": 79,
            "type": "ObjectExpression",
          },
        ],
        "callee": {
          "end": 78,
          "name": "Page",
          "start": 74,
          "type": "Identifier",
        },
        "end": 249,
        "optional": false,
        "start": 74,
        "type": "CallExpression",
      },
      "start": 74,
      "type": "ExpressionStatement",
    },
  ],
  "end": 250,
  "sourceType": "module",
  "start": 0,
  "type": "Program",
}
`;

exports[`require > should 1 1`] = `
[
  {
    "end": 33,
    "start": 25,
    "value": "./part",
  },
  {
    "end": 71,
    "start": 62,
    "value": "./other",
  },
  {
    "async": true,
    "end": 152,
    "start": 143,
    "value": "./async",
  },
  {
    "end": 199,
    "start": 189,
    "value": "./async2",
  },
]
`;

exports[`require > should 1 2`] = `
"const { part } = require('./part')

const { other } = require('./other')

Page({
  async onLoad() {
    const mod = await import('./async')
    
    const mod2 = await import('./async2')

    console.log(part, other, mod, mod2)
  },
})
"
`;

exports[`require > should 1 3`] = `"'./part'"`;

exports[`require > should 1 4`] = `"'./other'"`;

exports[`require > should 1 5`] = `"'./async'"`;

exports[`require > should 1 6`] = `"'./async2'"`;
